src/gdb_hostio.c

```diff
@@ -73,8 +73,16 @@ int hostio_reply(target_controller_s *const tc, char *const pbuf, const int len)
 static int hostio_get_response(target_controller_s *const tc)
 {
 	char *const packet_buffer = gdb_packet_buffer();
-	const size_t size = gdb_getpacket(packet_buffer, GDB_PACKET_BUFFER_SIZE);
-	return gdb_main_loop(tc, packet_buffer, GDB_PACKET_BUFFER_SIZE, size, true);
+	/* Still have to service normal 'X'/'m'-packets */
+	while (true) {
+		const size_t size = gdb_getpacket(packet_buffer, GDB_PACKET_BUFFER_SIZE);
+		/* GDB will need us to proxy access the inferior's memory */
+		if (packet_buffer[0] != 'F')
+			gdb_main_loop(tc, packet_buffer, GDB_PACKET_BUFFER_SIZE, size, false);
+		else
+			/* GDB finished proxied inferior memory manipulations and returns syscall exitcode */
+			return gdb_main_loop(tc, packet_buffer, GDB_PACKET_BUFFER_SIZE, size, true);
+	};
```

Comment:
Given the error handling in gdb_main_loop() and the insensitivity to the in_syscall parameter for the entire rest of the handling switch-case (it only impacts `F` packets), this feels like it would be better pivoted slightly. It also needs to handle the BMDA shutdown command otherwise that will cause an infinite loop here:

```suggestion
	while (true) {
		/* Get back the next packet to process and have the main loop handle it */
		const size_t size = gdb_getpacket(packet_buffer, GDB_PACKET_BUFFER_SIZE);
#if PC_HOSTED == 1
		/* If the packet was a escape packet and we're supposed to be shutting down, exit */
		if (size == 1U && packet_buffer[0] == '\x04' && shutdown_bmda)
			return 0;
#endif
		const int result = gdb_main_loop(tc, packet_buffer, GDB_PACKET_BUFFER_SIZE, size, true);
		/* If this was an F-packet, we're done */
		if (packet_buffer[0] == 'F')
			return result;
	}
```

Note that you have a trailing `;` on the loop which we're surprised is not linting.

Reply:
Adjusted to suggested-changes. I disagree on PC_HOSTED because this code is unreachable under BMDA (it never goes here but uses desktop OS syscalls directly).

Reply:
Ah, right, because it handles those requests an entirely different way - sure, you can drop that chunk of the suggestion then.

Reply:
Though I wonder if Ctrl+D or \x04 event handling should be added here in the rare race condition event that the user presses this in the short timeframe of a syscall in flight (or does gdb guard against this? returning -1 and getting us back to first main loop)

Reply:
Unsure - it's entirely possible that GDB just slams the link closed, at which point all gdb_getpacket() will ever return are injected `'\x04'` packets making this forever busy loop - which is the essence of the worry we had in the case of `mon shutdown` too for BMDA. It's a narrow window when this could happen, but perhaps err on the side of caution?

Reply:
My logic is as follows: in case gdb_getpacket() indeed returns a genuine (or injected by DTR-low/link loss) `\x04`, let's fail the semihosting call with a -1, make BMP resume the target and give it a chance to see this. Then the very next gdb_main_loop() called from bmp_poll_loop() will trigger on another `\x04` and detach from the target.
The other option is not to do this and let gdb_main_loop() detach in my face, but then I'd need to `if(!cur_target) return 0;`

Reply:
That logic seems sound to us 👍🏼 let's do that then.